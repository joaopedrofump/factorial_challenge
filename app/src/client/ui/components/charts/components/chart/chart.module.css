.container {
  padding: 1rem;
  background-color: var(--paper);
  box-shadow: 0 3px 10px rgb(0 0 0 / 0.2);
  border-radius: 7px;
  margin: 1rem; 
}

.chartWrapper {
  display: flex;
  justify-content: center;
  align-items: center;
}

.title {
  font-size: 1rem; 
  font-weight: lighter;
  margin: 0;
}

.chartHeader {
  display: flex;
  justify-content: space-between;
  margin-bottom: 1rem;
}

.means {
  padding: 0;
  list-style: none;
  font-size: small;
  font-weight: 200;
  margin: 0;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  min-height: 50px;
}

.means li {
  display: flex;
  align-items: center;
  position: relative;
}

.dot {
  padding: 0 3px;
  height: 2px;
  background-color: var(--secondary-color);
  width: 20px;
  margin-right: 5px;
  display: flex;
  align-items: center;
}

.dotMin {
  justify-content: flex-end;
}

.dotHour {
  justify-content: center;
}

.dotDay {
  justify-content: flex-start;
}

.dot > div {
  width: 6px;
  height: 6px;
  border-radius: 50%;
  background-color: var(--secondary-color);
}

.means li:hover {
  cursor: pointer;
}

.means li:hover > .popover {
  display: flex;
}

.popover {
  display: none;
  position: absolute;
  height: 45px;
  padding: 1rem;
  width: 150px;
  color: #fff;
  background-color: var(--primary-color);
  box-shadow: 0 3px 10px rgb(0 0 0 / 0.5);
  justify-content: center;
  align-items: center;
  right: 60px;
  border-radius: 7px;
}



.container *:focus {
  outline: none
}