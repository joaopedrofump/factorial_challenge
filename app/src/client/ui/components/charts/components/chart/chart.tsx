import { FC } from "react";
import {
  Area,
  AreaChart, ResponsiveContainer,
  Tooltip
} from "recharts";
import { theme } from "src/client/theme";
import { Form } from "src/client/ui/components/form/form";
import { Measure, Measurement } from "src/common/model/measure";
import styles from "./chart.module.css";
import { CustomTooltip } from "./components/custom-tooltip";

type Props = {
  measurement: Measurement;
  index: number;
  updateMeasures: (measures: Measurement) => void;
};

const Chart: FC<Props> = ({
  measurement: { measures, name, mean },
  index,
  updateMeasures,
}) => {
  const color =
    index % 2 === 0 ? theme.colors.primaryColor : theme.colors.secondaryColor;

  return (
    <article id="app-title" className={styles.container}>
      <div className={styles.chartHeader}>
        <h3 className={styles.title}>{name}</h3>
        <div>
          {mean && (
            <ul className={styles.means}>
              {mean.lastMin && (
                <li>
                   <div className={styles.popover}>
                      Last Minute Mean
                    </div>
                  <div className={`${styles.dot} ${styles.dotMin}`}>
                    <div />
                  </div>
                  <span>{mean.lastMin.toFixed(1)}</span>
                </li>
              )}
              {mean.lastHour && (
                <li>
                   <div className={styles.popover}>
                      Last Hour Mean
                    </div>
                  <div className={`${styles.dot} ${styles.dotHour}`}>
                    <div />
                  </div>
                  <span>{mean.lastHour.toFixed(1)}</span>
                </li>
              )}
              {mean.lastDay && (
                <li>
                   <div className={styles.popover}>
                      Last Day Mean
                    </div>
                  <div className={`${styles.dot} ${styles.dotDay}`}>
                    <div />
                  </div>
                  <span>{mean.lastDay.toFixed(1)}</span>
                </li>
              )}
            </ul>
          )}
        </div>
      </div>
      <div className={styles.chartWrapper}>
        <ResponsiveContainer width="95%" height={200}>
          <AreaChart
            data={measures}
            margin={{
              top: 0,
              right: 7,
              left: 7,
              bottom: 0,
            }}
          >
            <Tooltip content={<CustomTooltip />} />
            <Area type="natural" dataKey="value" strokeWidth={0} fill={color} />
          </AreaChart>
        </ResponsiveContainer>
      </div>
      <div className={styles.chartFooter}>
        <Form
          updateMeasures={updateMeasures}
          hcMeasurement={name}
          valuePlaceHolder="add more"
        />
      </div>
    </article>
  );
};



export { Chart };
