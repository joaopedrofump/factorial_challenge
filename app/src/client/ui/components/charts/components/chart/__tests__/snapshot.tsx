import { render } from "@testing-library/react";
import { Chart } from "src/client/ui/components/charts/components/chart/chart";

const mockedForm = jest.fn();

jest.mock("src/client/ui/components/form/form", () => ({
  Form: (args: unknown) => mockedForm(args),
}));

jest.mock("recharts", () => ({
  Area: (args: unknown) => <div />,
  AreaChart: (args: unknown) => <div />,
  ResponsiveContainer: (args: unknown) => <div />,
  Tooltip: (args: unknown) => <div />,
}));

it("renders a chart", () => {
  const measurement = {
    name: "m1",
    mean: { lastDay: 11, lastHour: 12, lastMin: 7 },
    measures: [
      { measurement: "m1", value: 12, timestamp: "t1" },
      { measurement: "m1", value: 12, timestamp: "t2" },
      { measurement: "m1", value: 12, timestamp: "t3" },
      { measurement: "m1", value: 12, timestamp: "t4" },
    ],
  };
  const { container } = render(
    <Chart index={0} measurement={measurement} updateMeasures={jest.fn()} />
  );
  expect(container).toMatchSnapshot();
});
