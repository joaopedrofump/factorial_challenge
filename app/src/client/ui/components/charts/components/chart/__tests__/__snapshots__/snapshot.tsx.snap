// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a chart 1`] = `
<div>
  <article
    class="container"
    id="app-title"
  >
    <div
      class="chartHeader"
    >
      <h3
        class="title"
      >
        m1
      </h3>
      <div>
        <ul
          class="means"
        >
          <li>
            <div
              class="popover"
            >
              Last Minute Mean
            </div>
            <div
              class="dot dotMin"
            >
              <div />
            </div>
            <span>
              7.0
            </span>
          </li>
          <li>
            <div
              class="popover"
            >
              Last Hour Mean
            </div>
            <div
              class="dot dotHour"
            >
              <div />
            </div>
            <span>
              12.0
            </span>
          </li>
          <li>
            <div
              class="popover"
            >
              Last Day Mean
            </div>
            <div
              class="dot dotDay"
            >
              <div />
            </div>
            <span>
              11.0
            </span>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="chartWrapper"
    >
      <div />
    </div>
    <div
      class="chartFooter"
    />
  </article>
</div>
`;
